Could not connect to rsync container. See build/README.md for setting up remote Docker engine.

I have symptoms that look like those being fixed by this: https://github.com/kubernetes/kubernetes/pull/34288 but I still get the error after cherry-picking that fix.

The last two commits in my tree are:

* 15b9583a713d4fd81aa446cceb224b94cfc62d3a (aka https://github.com/kubernetes/kubernetes/pull/34257)
* a cherry pick of https://github.com/kubernetes/kubernetes/pull/34288

I have removed **all** containers and images from my system. I'm in a clean git tree.

I run `make clean; make release` and get the following out:

```shell
+++ [1006 20:34:06] Verifying Prerequisites....
+++ [1006 20:34:06] Building Docker image kube-build:build-be4685d15b-4-v1.6.3-9
+++ [1006 20:35:42] Creating data container kube-build-data-be4685d15b-4-v1.6.3-9                                                                                                                               +++ [1006 20:35:44] Syncing sources to container
!!! [1006 20:35:49] Could not connect to rsync container. See build/README.md for setting up remote Docker engine.
make: *** [Makefile:239: release] Error 1
```

When I try to investigate...

```shell
$ docker ps -a

CONTAINER ID        IMAGE                                    COMMAND                  CREATED              STATUS                      PORTS                       NAMES
43529d54bfce        kube-build:build-be4685d15b-4-v1.6.3-9   "/rsyncd.sh"             58 seconds ago       Up 57 seconds               127.0.0.1:32776->8730/tcp   kube-rsync-be4685d15b-4-v1.6.3-9
e18d40dbcba6        kube-build:build-be4685d15b-4-v1.6.3-9   "chown -R 1000.1000 /"   About a minute ago   Exited (0) 58 seconds ago                               kube-build-data-be4685d15b-4-v1.6.3-9

$ docker logs 435
2016/10/06 20:35:45 [1] rsyncd version 3.1.1 starting, listening on port 8730
```

I'm happy to try anything for the sake of troubleshooting/debugging. Thanks.

It seems like I was able to connect... though my container was listening on 32777 instead.

```shell
$ make release
+++ [1006 23:10:42] Verifying Prerequisites....
+++ [1006 23:10:42] Syncing sources to container
!!! [1006 23:10:47] Could not connect to rsync container. See build/README.md for setting up remote Docker engine.
make: *** [Makefile:239: release] Error 1                                                                                                                                                                        

$ docker ps
CONTAINER ID        IMAGE                                    COMMAND             CREATED             STATUS              PORTS                       NAMES
0feebecbdd03        kube-build:build-be4685d15b-4-v1.6.3-9   "/rsyncd.sh"        3 minutes ago       Up 3 minutes        127.0.0.1:32777->8730/tcp   kube-rsync-be4685d15b-4-v1.6.3-9

$ telnet localhost 32777
Connection failed: Connection refused
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
@RSYNCD: 31.0
```

How did you run that function "in isolation"? I wrote out a file, and tried to execute it:

```shell
$ cd build

$ cat ./please-work.sh
#!/bin/bash
source common.sh
kube::build::rsync_probe 127.0.0.1 32777

$ ./please-work.sh
common.sh: line 589: LOCAL_OUTPUT_BUILD_CONTEXT: unbound variable
```

--- a/build/common.sh
+++ b/build/common.sh
@@ -147,6 +147,7 @@ kube::build::get_docker_wrapped_binaries() {
 function kube::build::verify_prereqs() {
   kube::log::status "Verifying Prerequisites...."
   kube::build::ensure_tar || return 1
+  kube::build::ensure_rsync || return 1
   kube::build::ensure_docker_in_path || return 1
   if kube::build::is_osx; then
       kube::build::docker_available_on_osx || return 1
@@ -233,6 +234,13 @@ function kube::build::is_osx() {
 
 function kube::build::is_gnu_sed() {
   [[ $(sed --version 2>&1) == *GNU* ]]
+}
+
+function kube::build::ensure_rsync() {
+  if [[ -z "$(which rsync)" ]]; then
+    kube::log::error "Can't find 'rsync' in PATH, please fix and retry."
+    return 1
+  fi
 }
 
 function kube::build::update_dockerfile() {

